alacritty/src/renderer/mod.rs

```diff
@@ -930,44 +932,50 @@ impl<'a, C> RenderApi<'a, C> {
         }
     }
 
-    fn render_batch(&mut self) {
+    fn render_batch(&mut self, layer_type: LayerType) {
+        let layer = &mut self.layers[layer_type as usize];
+
         unsafe {
             gl::BufferSubData(
                 gl::ARRAY_BUFFER,
                 0,
-                self.batch.size() as isize,
-                self.batch.instances.as_ptr() as *const _,
+                layer.batch.size() as isize,
+                layer.batch.instances.as_ptr() as *const _,
             );
         }
 
         // Bind texture if necessary.
-        if *self.active_tex != self.batch.tex {
+        if self.bound_tex != layer.batch.tex {
             unsafe {
-                gl::BindTexture(gl::TEXTURE_2D, self.batch.tex);
+                gl::BindTexture(gl::TEXTURE_2D, layer.batch.tex);
             }
-            *self.active_tex = self.batch.tex;
+            self.bound_tex = layer.batch.tex;
+            layer.active_tex = layer.batch.tex;
         }
 
         unsafe {
-            self.program.set_background_pass(true);
+            self.program.set_pass(0);
             gl::DrawElementsInstanced(
                 gl::TRIANGLES,
                 6,
                 gl::UNSIGNED_INT,
                 ptr::null(),
-                self.batch.len() as GLsizei,
+                layer.batch.len() as GLsizei,
             );
-            self.program.set_background_pass(false);
+            self.program.set_pass(match &layer_type {
+                LayerType::ColoredGlyphs => 2,
+                _ => 1,
+            });
```

Comment:
Maybe set_pass should take a `LayerType`? Creating a Background layer is probably a good idea anyways. Then just assign the fixed numbers to the enum and convert it to a number so we don't need to match it.